pkg/argo_client/manifests.go

```diff
@@ -1,116 +1,439 @@
+		// only if there isn't already an outgoing error.
+		if c := w.Close(); err == nil {
+			err = c
+		}
+	}()
+
+	_, err = io.Copy(w, r)
+	return err
+}
+
+func packageApp(ctx context.Context, source v1alpha1.ApplicationSource, refs []v1alpha1.ApplicationSource, repo *git.Repo, getRepo getRepo) (string, error) {
+	tempDir, err := os.MkdirTemp("", "package-*")
+	if err != nil {
+		return "", errors.Wrap(err, "failed to make temp dir")
+	}
+
+	tempAppDir := filepath.Join(tempDir, source.Path)
+	appPath := filepath.Join(repo.Directory, source.Path)
+
+	// copy app files to the temp dir
+	if err = filepath.Walk(appPath, func(path string, info fs.FileInfo, err error) error {
 		if err != nil {
-			log.Warn().Err(err).Msg("Failed to format manifest")
-			continue
+			return err
 		}
-		manifests = append(manifests, fmt.Sprintf("---\n%s", string(ret)))
+
+		if info.IsDir() {
+			return nil
+		}
+
+		relPath, err := filepath.Rel(appPath, path)
+		if err != nil {
+			return errors.Wrapf(err, "failed to calculate rel between %q and %q", appPath, path)
+		}
+		src := path
+		dst := filepath.Join(tempAppDir, relPath)
+		if err := copyFile(src, dst); err != nil {
+			return errors.Wrapf(err, "failed to %s => %s", src, dst)
+		}
+		return nil
+	}); err != nil {
+		return "", errors.Wrap(err, "failed to copy files")
 	}
-	return manifests
+
+	if source.Helm != nil {
+		refsByName := make(map[string]v1alpha1.ApplicationSource)
+		for _, ref := range refs {
+			refsByName[ref.Ref] = ref
+		}
+
+		for index, valueFile := range source.Helm.ValueFiles {
+			if strings.HasPrefix(valueFile, "$") {
+				refName, refPath, err := splitRefFromPath(valueFile)
+				if err != nil {
+					return "", errors.Wrap(err, "failed to parse value file")
+				}
+
+				ref, ok := refsByName[refName]
+				if !ok {
+					return "", errors.Wrap(err, "value file points at missing ref")
+				}
+
+				refRepo := repo
+				if !areSameRepos(ref.RepoURL, repo.CloneURL) {
+					refRepo, err = getRepo(ctx, ref.RepoURL, ref.TargetRevision)
+					if err != nil {
+						return "", errors.Wrapf(err, "failed to clone repo: %q", ref.RepoURL)
+					}
+				}
+
+				src := filepath.Join(refRepo.Directory, refPath)
+				dst := filepath.Join(tempDir, refPath)
+				if err = copyFile(src, dst); err != nil {
+					// handle source.spec.helm.ignoreMissingValues = true
+					if errors.Is(err, os.ErrNotExist) && source.Helm.IgnoreMissingValueFiles {
+						log.Debug().Str("valueFile", valueFile).Msg("ignore missing values file, because source.Helm.IgnoreMissingValueFiles is true")
+					} else {
+						return "", errors.Wrapf(err, "failed to copy referenced value file: %q", valueFile)
+					}
+				}
+
+				relPath, err := filepath.Rel(tempAppDir, dst)
+				if err != nil {
+					return "", errors.Wrap(err, "failed to find a relative path")
+				}
+				source.Helm.ValueFiles[index] = relPath
+				continue
+			}
+
+			relPath, err := filepath.Rel(source.Path, valueFile)
```

Comment:
This is all very new to me! Are you talking about the [helm-secrets](https://github.com/jkroepke/helm-secrets) plugin? Yeah, I suspect this would be quite broken, I'll have to think about how to do this =/

Reply:
yeah! That was my main usecase for starting the server-side integration. I was using argocd and helm secrets together. 

https://github.com/jkroepke/helm-secrets/blob/main/docs/ArgoCD%20Integration.md#external-chart-and-local-values

Reply:
The likely way this will work is if we packaged the whole repo and not just the modified app. That way, we maintain the directory structure and files in it